test/test_jit.py

```diff
@@ -457,6 +457,52 @@ def method(self, x):
         finally:
             torch.jit._enabled = True
 
+    def test_train_eval(self):
+        class Sub(nn.Module):
+            def forward(self, input):
+                if self.training:
+                    return input
+                else:
+                    return -input
+
+        class MyModule(torch.jit.ScriptModule):
+            def __init__(self):
+                super(MyModule, self).__init__()
+                self.sub = Sub()
+
+            @torch.jit.script_method
+            def forward(self, input):
+                return self.sub(input) + 1
+
+        m = MyModule()
+        input = torch.rand(3, 4)
+        self.assertEqual(input + 1, m(input))
+        m.eval()
+        self.assertEqual(-input + 1, m(input))
+
+    def test_train_eval_const(self):
+        class MyModule(torch.jit.ScriptModule):
+            __constants__ = ['training']
+
+            def __init__(self):
+                super(MyModule, self).__init__()
+                # TODO: it is illegal to try to call
+                # eval/train because training has already
+                # been set. Consider allowing
+                # constants to be mutable until the end of __init__
```

Comment:
Well if that's the case then having it be a constant is a bit useless. Pretty much every `not self.training` branch is guaranteed to always be dead code.

Reply:
Yes, it is useless, but I wanted to make sure it was consistent with how constants behaved at least.  In the future, we should support non-tensor module parameters so that training can be an actual parameter.